File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/VerticalStack.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.InterfaceBaseItem;
import com.spvessel.spacevil.Core.InterfaceVLayout;
import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Flags.ItemAlignment;
import com.spvessel.spacevil.Flags.SizePolicy;

public class VerticalStack extends Prototype implements InterfaceVLayout {
    private static int count = 0;

    private ItemAlignment _contentAlignment = ItemAlignment.TOP;

    public void setContentAlignment(ItemAlignment alignment) {
        if (alignment == ItemAlignment.TOP || alignment == ItemAlignment.VCENTER || alignment == ItemAlignment.BOTTOM)
            _contentAlignment = alignment;
    }

    public ItemAlignment getContentAlignment() {
        return _contentAlignment;
    }

    /**
     * Constructs a VerticalStack
     */
    public VerticalStack() {
        setItemName("verticalStack_" + count);
        count++;
        // setStyle(DefaultsService.getDefaultStyle("SpaceVIL.VerticalStack"));
        setStyle(DefaultsService.getDefaultStyle(VerticalStack.class));
        isFocusable = false;
    }

    // overrides
    @Override
    protected boolean getHoverVerification(float xpos, float ypos) {
        return false;
    }

    /**
     * Add item to the VerticalStack
     */
    @Override
    public void addItem(InterfaceBaseItem item) {
        super.addItem(item);
        updateLayout();
    }

    @Override
    public void insertItem(InterfaceBaseItem item, int index) {
        super.insertItem(item, index);
        updateLayout();
    }

    @Override
    public boolean removeItem(InterfaceBaseItem item) {
        boolean result = super.removeItem(item);
        if (result)
            updateLayout();
        return result;
    }

    /**
     * Set height of the VerticalStack
     */
    @Override
    public void setHeight(int height) {
        super.setHeight(height);
        updateLayout();
    }

    /**
     * Set Y position of the VerticalStack
     */
    @Override
    public void setY(int _y) {
        super.setY(_y);
        updateLayout();
    }

    /**
     * Update all children and VerticalStack sizes and positions according to
     * confines
     */
    public void updateLayout() {
        int total_space = getHeight() - getPadding().top - getPadding().bottom;
        int free_space = total_space;
        int fixed_count = 0;
        int expanded_count = 0;

        for (InterfaceBaseItem child : getItems()) {
            if (child.isVisible()) {
                if (child.getHeightPolicy() == SizePolicy.FIXED) {
                    fixed_count++;
                    free_space -= (child.getHeight() + child.getMargin().top + child.getMargin().bottom);//
                } else {
                    expanded_count++;
                }
            }
        }
        free_space -= getSpacing().vertical * ((expanded_count + fixed_count) - 1);

        int height_for_expanded = 1;
        if (expanded_count > 0 && free_space > expanded_count)
            height_for_expanded = free_space / expanded_count;

        int offset = 0;
        int startY = getY() + getPadding().top;

        if (expanded_count > 0 || _contentAlignment.equals(ItemAlignment.TOP)) {
            for (InterfaceBaseItem child : getItems()) {
                if (child.isVisible()) {
                    child.setY(startY + offset + child.getMargin().top);//
                    if (child.getHeightPolicy() == SizePolicy.EXPAND) {
                        if (height_for_expanded - child.getMargin().top - child.getMargin().bottom < child
                                .getMaxHeight())//
                            child.setHeight(height_for_expanded - child.getMargin().top - child.getMargin().bottom);//
                        else {
                            expanded_count--;
                            free_space -= (child.getHeight() + child.getMargin().top + child.getMargin().bottom);
                            height_for_expanded = 1;
                            if (expanded_count > 0 && free_space > expanded_count)
                                height_for_expanded = free_space / expanded_count;
                        }
                    }
                    offset += child.getHeight() + getSpacing().vertical + child.getMargin().top
                            + child.getMargin().bottom;//
                }
                // refactor
                child.setConfines();
            }
        } else {
            if (_contentAlignment.equals(ItemAlignment.BOTTOM)) {
                for (InterfaceBaseItem child : getItems()) {
                    if (child.isVisible()) {
                        child.setY(startY + offset + child.getMargin().top + free_space);//
                        offset += child.getHeight() + getSpacing().vertical + child.getMargin().top
                                + child.getMargin().bottom;//
                    }
                    child.setConfines();
                }
            } else if (_contentAlignment.equals(ItemAlignment.VCENTER)) {
                for (InterfaceBaseItem child : getItems()) {
                    if (child.isVisible()) {
                        child.setY(startY + offset + child.getMargin().top + free_space / 2);//
                        offset += child.getHeight() + getSpacing().vertical + child.getMargin().top
                                + child.getMargin().bottom;//
                    }
                    child.setConfines();
                }
            }
        }

    }
}

File: spacevil_21052019/net_src/Items/VerticalStack.cs
using System;
using System.Drawing;
using SpaceVIL.Common;
using SpaceVIL.Core;
using System.Collections.Generic;

namespace SpaceVIL
{
    public class VerticalStack : Prototype, IVLayout
    {
        static int count = 0;

        private ItemAlignment _contentAlignment = ItemAlignment.Top;

        public void SetContentAlignment(ItemAlignment alignment)
        {
            if (alignment == ItemAlignment.Top || alignment == ItemAlignment.VCenter || alignment == ItemAlignment.Bottom)
                _contentAlignment = alignment;
        }

        public ItemAlignment GetContentAlignment()
        {
            return _contentAlignment;
        }

        public VerticalStack()
        {
            SetItemName("VerticalStack_" + count);
            count++;
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.VerticalStack)));
            IsFocusable = false;

        }

        //overrides
        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            return false;
        }

        public override void AddItem(IBaseItem item)
        {
            base.AddItem(item);
            UpdateLayout();
        }
        public override void InsertItem(IBaseItem item, int index)
        {
            base.InsertItem(item, index);
            UpdateLayout();
        }

        public override bool RemoveItem(IBaseItem item)
        {
            bool result = base.RemoveItem(item);
            if (result)
                UpdateLayout();
            return result;
        }
        public override void SetHeight(int height)
        {
            base.SetHeight(height);
            UpdateLayout();
        }
        public override void SetY(int _y)
        {
            base.SetY(_y);
            UpdateLayout();
        }

        public void UpdateLayout()
        {
            int total_space = GetHeight() - GetPadding().Top - GetPadding().Bottom;
            int free_space = total_space;
            int fixed_count = 0;
            int expanded_count = 0;

            List<int> maxHeightExpands = new List<int>();

            foreach (var child in GetItems())
            {
                if (child.IsVisible())
                {
                    if (child.GetHeightPolicy() == SizePolicy.Fixed)
                    {
                        fixed_count++;
                        free_space -= (child.GetHeight() + child.GetMargin().Top + child.GetMargin().Bottom);//
                    }
                    else
                    {
                        if (child.GetMaxHeight() < total_space) {
                            maxHeightExpands.Add(child.GetMaxHeight() + child.GetMargin().Top + child.GetMargin().Bottom);
                        }
                        expanded_count++;
                    }
                }
            }
            free_space -= GetSpacing().Vertical * ((expanded_count + fixed_count) - 1);

            int height_for_expanded = 1;
            // if (expanded_count > 0 && free_space > expanded_count)
            //     height_for_expanded = free_space / expanded_count;

            maxHeightExpands.Sort();

            while (true) {
                if (expanded_count == 0)
                    break;

                if (free_space > expanded_count)
                    height_for_expanded = free_space / expanded_count;

                if (height_for_expanded <= 1 || maxHeightExpands.Count == 0) {
                    break;
                }

                if (height_for_expanded > maxHeightExpands[0]) {
                    while (maxHeightExpands.Count > 0 && height_for_expanded > maxHeightExpands[0]) {
                        free_space -= maxHeightExpands[0];
                        maxHeightExpands.RemoveAt(0);
                        expanded_count--;
                    }
                } else {
                    break;
                }
            }

            int offset = 0;
            int startY = GetY() + GetPadding().Top;
            bool isFirstExpand = false;
            int diff = (free_space - height_for_expanded * expanded_count);
            if (expanded_count != 0 && diff > 0) {
                isFirstExpand = true;
            }

            if (expanded_count > 0 || _contentAlignment.Equals(ItemAlignment.Top))
            {
                foreach (var child in GetItems())
                {
                    if (child.IsVisible())
                    {
                        child.SetY(startY + offset + child.GetMargin().Top);//
                        if (child.GetHeightPolicy() == SizePolicy.Expand)
                        {
                            if (height_for_expanded - child.GetMargin().Top - child.GetMargin().Bottom < child.GetMaxHeight())//
                            {
                                child.SetHeight(height_for_expanded - child.GetMargin().Top - child.GetMargin().Bottom);//
                            }
                            else
                            {
                                // expanded_count--;
                                // free_space -= (child.GetHeight() + child.GetMargin().Top + child.GetMargin().Bottom);
                                // height_for_expanded = 1;
                                // if (expanded_count > 0 && free_space > expanded_count)
                                //     height_for_expanded = free_space / expanded_count;
                                child.SetHeight(child.GetMaxHeight()); //?
                            }

                            if (isFirstExpand) {
                                if (child.GetHeight() + diff < child.GetMaxHeight()) {
                                    child.SetHeight(child.GetHeight() + diff);
                                    isFirstExpand = false;
                                }
                            }
                        }
                        offset += child.GetHeight() + GetSpacing().Vertical + child.GetMargin().Top + child.GetMargin().Bottom;//
                    }
                    //refactor
                    child.SetConfines();
                }
            }
            else
            {
                //for content alignment right
                if (_contentAlignment.Equals(ItemAlignment.Bottom))
                {
                    foreach (var child in GetItems())
                    {
                        if (child.IsVisible())
                        {
                            child.SetY(startY + offset + child.GetMargin().Top + free_space);
                            offset += child.GetHeight() + GetSpacing().Vertical + child.GetMargin().Top + child.GetMargin().Bottom;//

                            if (child.GetHeightPolicy() == SizePolicy.Expand)
                                child.SetHeight(child.GetMaxHeight());
                        }
                        child.SetConfines();
                    }
                }
                //for content alignment hcenter
                else if (_contentAlignment.Equals(ItemAlignment.VCenter))
                {
                    foreach (var child in GetItems())
                    {
                        if (child.IsVisible())
                        {
                            child.SetY(startY + offset + child.GetMargin().Top + free_space / 2);
                            offset += child.GetHeight() + GetSpacing().Vertical + child.GetMargin().Top + child.GetMargin().Bottom;//

                            if (child.GetHeightPolicy() == SizePolicy.Expand)
                                child.SetHeight(child.GetMaxHeight());
                        }
                        child.SetConfines();
                    }
                }
            }
        }
    }
}